### PR TITLE
Support ViTDet algo

### PR DESCRIPTION
**Pull request by tuofeilunhifi**
1. support ViTDet algo
2. bugfix：fp16 error
3. bugfix：use large-scale jilttering，augmentation pipeline raise Nonetype

vitdet mask-rcnn: 50.6

### CODE REVIEW

**Comment by Cathy0908** [comment]
remove configs/detection/maskrcnn directory

File: easycv/datasets/detection/raw.py [hunk 1]
```diff
@@ -31,9 +31,13 @@ def __len__(self):
...
 
     def __getitem__(self, idx):
-        data_dict = self.data_source.get_sample(idx)
-        data_dict = self.pipeline(data_dict)
-        return data_dict
+        while True:
+            data_dict = self.data_source.get_sample(idx)
+            data_dict = self.pipeline(data_dict)
+            if data_dict is None:
+                idx = self.data_source._rand_another(idx)
```
Comment (Cathy0908) [inline easycv/datasets/detection/raw.py:38]: detection data source has no attribute  of `_rand_another`

File: easycv/models/detection/vitdet/rpn_head_norm.py [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+import torch.nn.functional as F
+from mmcv.cnn import ConvModule
+from mmcv.ops import batched_nms
+from mmdet.models import AnchorHead
+
+from easycv.models.registry import HEADS
+
+
+@HEADS.register_module()
+class RPNHead_Norm(AnchorHead):
```
Comment (Cathy0908) [inline easycv/models/detection/vitdet/rpn_head_norm.py:15]: modify `RPNHead_Norm` to `RPNHeadNorm`

File: easycv/utils/config_tools.py [hunk 3]
```diff
@@ -15,6 +15,14 @@
     import re
 
 
+def pre_check_config(cfg):
```
Comment (Cathy0908) [inline easycv/utils/config_tools.py:18]: remove `pre_check_config`

File: easycv/utils/fp16_utils.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
+from collections import abc
```
Comment (Cathy0908) [inline easycv/utils/fp16_utils.py:1]: add license

File: tools/test.py [hunk 5]
```diff
@@ -0,0 +1,145 @@
+# Copyright (c) Alibaba, Inc. and its affiliates.
```
Comment (Cathy0908) [inline tools/test.py:1]: remove test.py

File: tools/eval.py [hunk 6]
```diff
@@ -205,7 +205,7 @@ def main():
 
     assert 'eval_pipelines' in cfg, 'eval_pipelines is needed for testting'
     for eval_pipe in cfg.eval_pipelines:
-        eval_data = eval_pipe.data
+        eval_data = cfg.data.val
```
Comment (Cathy0908) [inline tools/eval.py:208]: `eval_data = eval_pipe.get('data', None) or cfg.data.val`

File: easycv/hooks/optimizer_hook.py [hunk 7]
```diff
@@ -103,11 +104,17 @@ def __init__(self,
         self.ignore_key_epoch = ignore_key_epoch
         if LooseVersion(torch.__version__) >= LooseVersion('1.6.0'):
             self.scaler = amp.GradScaler()
+            self._scale_update_param = 512.
 
     def before_run(self, runner):
-        runner.fp16_enable = True
+        """Preparing steps before Mixed Precision Training."""
+        # wrap model mode to fp16
+        wrap_fp16_model(runner.model)
```
Comment (Cathy0908) [inline easycv/hooks/optimizer_hook.py:112]: this is `AMPFP16OptimizerHook`, support mixed precision

but `wrap_fp16_model` use model.half() , convert all the model weights to half precision.

remove `wrap_fp16_model`

File: configs/detection/vitdet/_base_/datasets/coco.py [hunk 8]
```diff
@@ -0,0 +1,104 @@
+CLASSES = [
```
Comment (Cathy0908) [inline configs/detection/vitdet/_base_/datasets/coco.py:1]: rename` coco.py `to `coco_instance.py`

File: easycv/core/optimizer/adam.py [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+        else:
+            denom = (exp_avg_sq.sqrt() / math.sqrt(bias_correction2)).add_(eps)
+
+        step_size = lr / bias_correction1
+
+        param.addcdiv_(exp_avg, denom, value=-step_size)
+
+
+@OPTIMIZERS.register_module()
+class _AdamW(Optimizer):
```
Comment (Cathy0908) [inline easycv/core/optimizer/adam.py:50]: rename `_AdamW` to `AdamW`, and rewritre this api to inherit torch optimizer

pop original registry if name conflict

File: easycv/core/optimizer/lars.py [hunk 10]
```diff
@@ -1,5 +1,10 @@
 # Copyright (c) Alibaba, Inc. and its affiliates.
+import math
+from typing import List
+
 import torch
+from torch import Tensor
+from torch.optim import *  # noqa: F401,F403
```
Comment (Cathy0908) [inline easycv/core/optimizer/lars.py:7]: remove useless imports

File: docs/source/model_zoo_detection.md [hunk 11]
```diff
@@ -12,3 +12,9 @@ Pretrained on COCO2017 dataset.
...
+
+## ViTDet
+
+| Algorithm  | Config                                                       | bbox_mAP<sup>val<br/><sub>0.5:0.95</sub> | mask_mAP<sup>val<br/><sub>0.5:0.95</sub> | Download                                                     |
+| ---------- | ------------------------------------------------------------ | ------------------------ | --------------- | ------------------------------------------------------------ |
+| ViTDet_MaskRCNN    | [vitdet_maskrcnn](https://github.com/alibaba/EasyCV/tree/master/configs/detection/vitdet/vitdet_100e.py) | 50.57                   | 44.96          | [model](https://pai-vision-data-hz.oss-cn-zhangjiakou.aliyuncs.com/EasyCV/modelzoo/detection/vitdet/vit_base/vitdet_maskrcnn.pth) - [log](https://pai-vision-data-hz.oss-cn-zhangjiakou.aliyuncs.com/EasyCV/modelzoo/detection/vitdet/vit_base/vitdet_maskrcnn.log.json) |
```
Comment (Cathy0908) [inline docs/source/model_zoo_detection.md:20]: whether the file is set to be read publicly

File: configs/detection/vitdet/_base_/models/vitdet_fcos.py [hunk 12]
```diff
@@ -0,0 +1,69 @@
+# model settings
+
```
Comment (wenmengzhou) [inline configs/detection/vitdet/_base_/models/vitdet_fcos.py:2]: 这个文件暂时先不要提交

File: easycv/core/optimizer/__init__.py [hunk 13]
```diff
@@ -1,4 +1,12 @@
+from distutils.version import LooseVersion
+
+import torch
 from torch.optim import *
 
+from .builder import build_optimizer_constructor
```
Comment (wenmengzhou) [inline easycv/core/optimizer/__init__.py:6]: not used

File: docs/source/model_zoo_detection.md [hunk 14]
```diff
@@ -12,3 +12,9 @@ Pretrained on COCO2017 dataset.
...
 | YOLOX-tiny | [yolox_tiny_8xb16_300e_coco](https://github.com/alibaba/EasyCV/tree/master/configs/detection/yolox/yolox_tiny_8xb16_300e_coco.py) | 31.5                   | 49.2          | [model](http://pai-vision-data-hz.oss-cn-zhangjiakou.aliyuncs.com/EasyCV/modelzoo/detection/yolox_tiny_bs16_lr002/epoch_300.pth) - [log](http://pai-vision-data-hz.oss-cn-zhangjiakou.aliyuncs.com/EasyCV/modelzoo/detection/yolox_tiny_bs16_lr002/log.txt) |
 | YOLOX-nano | [yolox_nano_8xb16_300e_coco](https://github.com/alibaba/EasyCV/tree/master/configs/detection/yolox/yolox_tiny_8xb16_300e_coco.py) | 26.5                   | 42.6          | [model](http://pai-vision-data-hz.oss-cn-zhangjiakou.aliyuncs.com/EasyCV/modelzoo/detection/yolox_nano_bs16_lr002/epoch_300.pth) - [log](http://pai-vision-data-hz.oss-cn-zhangjiakou.aliyuncs.com/EasyCV/modelzoo/detection/yolox_nano_bs16_lr002/log.txt) |
+
+## ViTDet
+
+| Algorithm  | Config                                                       | bbox_mAP<sup>val<br/><sub>0.5:0.95</sub> | mask_mAP<sup>val<br/><sub>0.5:0.95</sub> | Download                                                     |
```
Comment (wenmengzhou) [inline docs/source/model_zoo_detection.md:18]: make the table header same as yolox

File: easycv/datasets/detection/raw.py [hunk 15]
```diff
@@ -31,9 +31,13 @@ def __len__(self):
         return self.num_samples
 
     def __getitem__(self, idx):
-        data_dict = self.data_source.get_sample(idx)
-        data_dict = self.pipeline(data_dict)
-        return data_dict
+        while True:
+            data_dict = self.data_source.get_sample(idx)
+            data_dict = self.pipeline(data_dict)
+            if data_dict is None:
```
Comment (wenmengzhou) [inline easycv/datasets/detection/raw.py:37]: print some log after a few times retry

File: easycv/utils/fp16_utils.py [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+import torch
+import torch.nn as nn
+from mmcv.utils import TORCH_VERSION, digit_version
+
+
+def cast_tensor_type(inputs, src_type, dst_type):
+    """Recursively convert Tensor in inputs from src_type to dst_type.
+
+    Note:
+        In v1.4.4 and later, ``cast_tersor_type`` will only convert the
```
Comment (wenmengzhou) [inline easycv/utils/fp16_utils.py:14]: can we directly use this function by importing from mmcv?

File: easycv/utils/fp16_utils.py [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+
+    Otherwise:
+    1. Convert FP32 model to FP16.
+    2. Remain some necessary layers to be FP32, e.g., normalization layers.
+    3. Set `fp16_enabled` flag inside the model to True.
+
+    Args:
+        model (nn.Module): Model in FP32.
+    """
+    if (TORCH_VERSION == 'parrots'
```
Comment (wenmengzhou) [inline easycv/utils/fp16_utils.py:69]: remove parrots condition

File: easycv/models/backbones/vitdet.py [hunk 18]
```diff
@@ -0,0 +1,849 @@
...
+from mmdet.utils import get_root_logger
+from timm.models.layers import drop_path, to_2tuple, trunc_normal_
+from torch.nn.modules.batchnorm import _BatchNorm
+
+from easycv.utils.checkpoint import load_checkpoint
+from ..registry import BACKBONES
+from ..utils import build_conv_layer
+
+
+class DropPath(nn.Module):
```
Comment (wenmengzhou) [inline easycv/models/backbones/vitdet.py:32]: NIT: duplicated code with easycv/models/backbones/vit_transfomer_dynamic.py, we should resuse these modules, add a TODO to refactor vision transformers in the future

File: easycv/models/backbones/vitdet.py [hunk 19]
```diff
@@ -0,0 +1,849 @@
+# --------------------------------------------------------
+# BEIT: BERT Pre-Training of Image Transformers (https://arxiv.org/abs/2106.08254)
```
Comment (wenmengzhou) [inline easycv/models/backbones/vitdet.py:2]: change copyright, keep the original copyright and add a link for original code,
add corresponding license in NOTICE file

**Review by wenmengzhou — COMMENTED** [review]
submit comment

File: easycv/models/detection/yolox/yolox.py [hunk 20]
```diff
@@ -151,6 +153,7 @@ def forward_test(self, img: Tensor, img_metas=None) -> Tensor:
 
         return test_outputs
 
+    @auto_fp16(apply_to=('img', ))
```
[the remark this answers is not in the document]
Comment (tuofeilunhifi) [reply easycv/models/detection/yolox/yolox.py:156]: base_model already modified

[the remark this answers is not in the document]
Comment (tuofeilunhifi) [reply easycv/models/detection/yolox/yolox.py:156]: > classification model has not been modified like this
base_model already modified

[the remark this answers is not in the document]
Comment (tuofeilunhifi) [reply easycv/models/detection/yolox/yolox.py:156]: > classification model has not been modified like this

base_model already modified

File: easycv/models/pose/top_down.py [hunk 21]
```diff
@@ -6,6 +6,7 @@
 import mmcv
 import numpy as np
 from mmcv.image import imwrite
+from mmcv.runner import auto_fp16
```
[the remark this answers is not in the document]
Comment (tuofeilunhifi) [reply easycv/models/pose/top_down.py:9]: > not used?

already modified

File: easycv/core/optimizer/layer_decay_optimizer_constructor.py [hunk 22]
```diff
@@ -0,0 +1,107 @@
+# Reference from https://github.com/ViTAE-Transformer/ViTDet/blob/main/mmcv_custom/layer_decay_optimizer_constructor.py
```
[the remark this answers is not in the document]
Comment (tuofeilunhifi) [reply easycv/core/optimizer/layer_decay_optimizer_constructor.py]: already remove file

File: easycv/models/detection/vitdet/sfp.py [hunk 23]
```diff
@@ -0,0 +1,245 @@
...
+
+    def forward(self, x):
+        x = x.permute(0, 2, 3, 1)
+        x = self.ln(x)
+        x = x.permute(0, 3, 1, 2).contiguous()
+        return x
+
+
+@NECKS.register_module()
+class SFP(BaseModule):
```
[the remark this answers is not in the document]
Comment (tuofeilunhifi) [reply easycv/models/detection/vitdet/sfp.py:25]: > FPN ?

simple feature pyramid

File: configs/detection/vitdet/_base_/datasets/coco.py (hunk 8 above)
Comment (tuofeilunhifi) [reply configs/detection/vitdet/_base_/datasets/coco.py]: already modified

File: easycv/core/optimizer/adam.py (hunk 9 above)
Comment (tuofeilunhifi) [reply easycv/core/optimizer/adam.py]: already modified

File: easycv/core/optimizer/lars.py (hunk 10 above)
Comment (tuofeilunhifi) [reply easycv/core/optimizer/lars.py:3]: already modified

File: easycv/datasets/detection/data_sources/base.py [hunk 24]
```diff
@@ -139,6 +139,9 @@ def post_process_fn(self, result_dict):
 
         return result_dict
 
+    def _rand_another(self, idx):
+        return (idx + 1) % self.num_samples
```
[the remark this answers is not in the document]
Comment (tuofeilunhifi) [reply easycv/datasets/detection/data_sources/base.py:143]: already modified

File: configs/detection/vitdet/_base_/models/vitdet_fcos.py (hunk 12 above)
Comment (tuofeilunhifi) [reply configs/detection/vitdet/_base_/models/vitdet_fcos.py]: already modified

File: easycv/core/optimizer/__init__.py (hunk 13 above)
Comment (tuofeilunhifi) [reply easycv/core/optimizer/__init__.py]: already modified

File: easycv/datasets/detection/raw.py (hunk 15 above)
Comment (tuofeilunhifi) [reply easycv/datasets/detection/raw.py:40]: already modified

File: easycv/utils/fp16_utils.py (hunk 16 above)
Comment (tuofeilunhifi) [reply easycv/utils/fp16_utils.py]: already modified

File: easycv/models/backbones/vitdet.py (hunk 18 above)
Comment (tuofeilunhifi) [reply easycv/models/backbones/vitdet.py:21]: already record todo

File: easycv/runner/ev_runner.py [hunk 25]
```diff
@@ -194,6 +183,9 @@ def load_checkpoint(self,
         Returns:
             dict or OrderedDict: The loaded checkpoint.
         """
+        if torch.cuda.is_available():
+            device_id = torch.cuda.current_device()
+            map_location = lambda storage, loc: storage.cuda(device_id)
```
[the remark this answers is not in the document]
Comment (tuofeilunhifi) [reply easycv/runner/ev_runner.py]: already remove

File: tests/tools/test_fp16.py [hunk 26]
```diff
@@ -0,0 +1,306 @@
...
+import unittest
+
+import numpy as np
+import torch
+import torch.nn as nn
+from mmcv.runner import auto_fp16, force_fp32
+from mmcv.runner.fp16_utils import cast_tensor_type
+
+
+class AutoFP16Test(unittest.TestCase):
```
[the remark this answers is not in the document]
Comment (tuofeilunhifi) [reply tests/tools/test_fp16.py:11]: add model_test_fp16

File: docs/source/model_zoo_detection.md (hunk 11 above)
Comment (tuofeilunhifi) [reply docs/source/model_zoo_detection.md:20]: already public

File: configs/detection/mask_rcnn/_base_/models/mask_rcnn_r50_fpn.py [hunk 27]
```diff
@@ -0,0 +1,138 @@
+# model settings
```
[the remark this answers is not in the document]
Comment (tuofeilunhifi) [reply configs/detection/mask_rcnn/_base_/models/mask_rcnn_r50_fpn.py]: already remove

File: configs/detection/vitdet/_base_/datasets/coco.py [hunk 28]
```diff
@@ -0,0 +1,104 @@
...
+    'bowl', 'banana', 'apple', 'sandwich', 'orange', 'broccoli', 'carrot',
+    'hot dog', 'pizza', 'donut', 'cake', 'chair', 'couch', 'potted plant',
+    'bed', 'dining table', 'toilet', 'tv', 'laptop', 'mouse', 'remote',
+    'keyboard', 'cell phone', 'microwave', 'oven', 'toaster', 'sink',
+    'refrigerator', 'book', 'clock', 'vase', 'scissors', 'teddy bear',
+    'hair drier', 'toothbrush'
+]
+
+# dataset settings
+data_root = '/apsarapangu/disk2/yunji.cjy/coco/'
```
[the remark this answers is not in the document]
Comment (tuofeilunhifi) [reply configs/detection/vitdet/_base_/datasets/coco.py]: already modified

File: configs/detection/vitdet/_base_/models/vitdet.py [hunk 29]
```diff
@@ -0,0 +1,149 @@
+# model settings
+# # Use MMSyncBN that handles empty tensor in head. It can be changed to
+# # SyncBN after https://github.com/pytorch/pytorch/issues/36530 is fixed
+# # Requires MMCV-full after  https://github.com/open-mmlab/mmcv/pull/1205.
+# norm_cfg = dict(type='MMSyncBN', requires_grad=True)
+
+norm_cfg = dict(type='GN', num_groups=1, requires_grad=True)
+
+pretrained = '/home/yunji.cjy/pretrain/warpper_mae_vit-base-p16-1600e.pth'
```
[the remark this answers is not in the document]
Comment (tuofeilunhifi) [reply configs/detection/vitdet/_base_/models/vitdet.py]: already modified

File: configs/detection/vitdet/_base_/models/vitdet.py [hunk 30]
```diff
@@ -0,0 +1,149 @@
+# model settings
+# # Use MMSyncBN that handles empty tensor in head. It can be changed to
+# # SyncBN after https://github.com/pytorch/pytorch/issues/36530 is fixed
+# # Requires MMCV-full after  https://github.com/open-mmlab/mmcv/pull/1205.
+# norm_cfg = dict(type='MMSyncBN', requires_grad=True)
```
[the remark this answers is not in the document]
Comment (tuofeilunhifi) [reply configs/detection/vitdet/_base_/models/vitdet.py]: already remove

File: configs/detection/vitdet/_base_/models/vitdet.py [hunk 31]
```diff
@@ -0,0 +1,149 @@
...
+            min_bbox_size=0),
+        rcnn=dict(
+            score_thr=0.05,
+            nms=dict(type='nms', iou_threshold=0.5),
+            max_per_img=100,
+            mask_thr_binary=0.5)))
+
+mmlab_modules = [
+    dict(type='mmdet', name='MaskRCNN', module='model'),
+    # dict(type=MMDET, name='ResNet', module='backbone'), # comment out, use EasyCV ResNet
```
[the remark this answers is not in the document]
Comment (tuofeilunhifi) [reply configs/detection/vitdet/_base_/models/vitdet.py]: already remove

File: easycv/apis/test.py [hunk 32]
```diff
@@ -213,7 +213,7 @@ def multi_gpu_test(model,
 
         if rank == 0:
             if 'img_metas' in data:
-                batch_size = len(data['img_metas'].data[0])
+                batch_size = len(data['img_metas'])
```
[the remark this answers is not in the document]
Comment (tuofeilunhifi) [reply easycv/apis/test.py]: already revert

File: easycv/core/optimizer/layer_decay_optimizer_constructor.py [hunk 33]
```diff
@@ -0,0 +1,105 @@
...
+        return 0
+    elif var_name.startswith('backbone.blocks'):
+        layer_id = int(var_name.split('.')[2])
+        return layer_id + 1
+    else:
+        return num_max_layer - 1
+
+
+@OPTIMIZER_BUILDERS.register_module()
+class LayerDecayOptimizerConstructor(DefaultOptimizerConstructor):
```
[the remark this answers is not in the document]
Comment (tuofeilunhifi) [reply easycv/core/optimizer/layer_decay_optimizer_constructor.py]: already remove

File: easycv/core/optimizer/optimizer.py [hunk 34]
```diff
@@ -0,0 +1,288 @@
...
+from typing import List
+
+import torch
+from mmcv.runner.optimizer.builder import OPTIMIZERS
+from torch import Tensor
+from torch.optim import *  # noqa: F401,F403
+from torch.optim.optimizer import Optimizer, required
+
+
+class LARS(Optimizer):
```
[the remark this answers is not in the document]
Comment (tuofeilunhifi) [reply easycv/core/optimizer/optimizer.py]: already split

File: easycv/datasets/detection/raw.py (hunk 1 above)
Comment (tuofeilunhifi) [reply easycv/datasets/detection/raw.py:42]: already modified

File: easycv/models/detection/vitdet/rpn_head_norm.py (hunk 2 above)
Comment (tuofeilunhifi) [reply easycv/models/detection/vitdet/rpn_head_norm.py]: already modified

File: easycv/utils/config_tools.py (hunk 3 above)
Comment (tuofeilunhifi) [reply easycv/utils/config_tools.py]: already remove

File: easycv/utils/fp16_utils.py (hunk 4 above)
Comment (tuofeilunhifi) [reply easycv/utils/fp16_utils.py]: already remove fp16_utils.py

File: tools/test.py (hunk 5 above)
Comment (tuofeilunhifi) [reply tools/test.py]: already remove

File: tools/eval.py (hunk 6 above)
Comment (tuofeilunhifi) [reply tools/eval.py]: already modified

File: easycv/hooks/optimizer_hook.py (hunk 7 above)
Comment (tuofeilunhifi) [reply easycv/hooks/optimizer_hook.py:112]: > 

uniform use mmcv fp16

File: easycv/models/backbones/vitdet.py (hunk 19 above)
Comment (tuofeilunhifi) [reply easycv/models/backbones/vitdet.py]: already change copyright

**Pull request merged**